SDIOBlockDevice.cpp

```diff
@@ -0,0 +1,378 @@
+    _block_size = BLOCK_SIZE_HC;
+    _erase_size = BLOCK_SIZE_HC;
+}
+
+SDIOBlockDevice::~SDIOBlockDevice()
+{
+    if (_is_initialized)
+    {
+        deinit();
+    }
+}
+
+int SDIOBlockDevice::init()
+{
+    debug_if(SD_DBG, "init Card...\r\n");
+
+    lock();
+
+    if (!_is_initialized)
+    {
+        _init_ref_count = 0;
+    }
+
+    _init_ref_count++;
+
+    if (_init_ref_count != 1)
+    {
+        unlock();
+        return BD_ERROR_OK;
+    }
+
+    if (isPresent() == false)
+    {
+        unlock();
+        return SD_BLOCK_DEVICE_ERROR_NO_DEVICE;
+    }
+
+    int status = SD_Init();
+    if (BD_ERROR_OK != status)
+    {
+        unlock();
+        return BD_ERROR_DEVICE_ERROR;
+    }
+
+    SD_GetCardInfo(&_cardInfo);
+    _is_initialized = true;
+    debug_if(SD_DBG, "SD initialized: type: %d  version: %d  class: %d\n",
+             _cardInfo.CardType, _cardInfo.CardVersion, _cardInfo.Class);
+    debug_if(SD_DBG, "SD size: %d MB\n",
+             _cardInfo.LogBlockNbr / 2 / 1024);
+
+    // get sectors count from cardinfo
+    _sectors = _cardInfo.LogBlockNbr;
+    if (BLOCK_SIZE_HC != _cardInfo.BlockSize)
+    {
+        unlock();
+        return SD_BLOCK_DEVICE_ERROR_UNSUPPORTED_BLOCKSIZE;
+    }
+
+    unlock();
+    return status;
+}
+
+int SDIOBlockDevice::deinit()
+{
+    debug_if(SD_DBG, "deinit Card...\r\n");
+    lock();
+
+    if (!_is_initialized)
+    {
+        _init_ref_count = 0;
+        unlock();
+        return BD_ERROR_OK;
+    }
+
+    _init_ref_count--;
+
+    if (_init_ref_count)
+    {
+        unlock();
+        return BD_ERROR_OK;
+    }
+
+    int status = SD_DeInit();
+    _is_initialized = false;
+
+    _sectors = 0;
+
+    unlock();
+    return status;
+}
+
+int SDIOBlockDevice::read(void *b, bd_addr_t addr, bd_size_t size)
+{
+    //debug_if(SD_DBG, "read Card...\r\n");
+    lock();
+    if (isPresent() == false)
+    {
+        unlock();
+        return SD_BLOCK_DEVICE_ERROR_NO_DEVICE;
+    }
+    if (!is_valid_read(addr, size))
+    {
+        unlock();
+        return SD_BLOCK_DEVICE_ERROR_PARAMETER;
+    }
+
+    if (!_is_initialized)
+    {
+        unlock();
+        return SD_BLOCK_DEVICE_ERROR_NO_INIT;
+    }
+
+    uint32_t *buffer = static_cast<uint32_t *>(b);
+
+    // ReadBlocks uses byte unit address
+    // SDHC and SDXC Cards different addressing is handled in ReadBlocks()
+    bd_addr_t blockCnt = size / _block_size;
+    addr = addr / _block_size;
+
+    // make sure card is ready
+    while (SD_GetCardState() != SD_TRANSFER_OK)
+    {
+        // wait until SD ready
+        wait_ms(1);
+    }
+
+    // receive the data : one block/ multiple blocks is handled in ReadBlocks()
+    int status = SD_ReadBlocks_DMA(buffer, addr, blockCnt);
+    debug_if(SD_DBG, "ReadBlocks dbgtest addr: %lld  blockCnt: %lld \n", addr, blockCnt);
+
+    if (status == MSD_OK)
+    {
+        // wait until DMA finished
+        while (SD_DMA_ReadPending() != SD_TRANSFER_OK)
```

Comment:
I tried first the polling version, but had lots of performance problems. The data fifo must be read/written very fast, otherwise you'll get over/underrun errors. Only the DMA was fast enough to avoid these problems.